compiler/src/org.graalvm.compiler.replacements/src/org/graalvm/compiler/replacements/StandardGraphBuilderPlugins.java

```diff
@@ -2282,4 +2283,25 @@ public SnippetTemplate.SnippetInfo getSnippet(BigIntegerSnippets.Templates templ
             }
         });
     }
+
+    private static boolean hasEnsureMaterializedForStackWalk() {
+        try {
+            Class.forName("java.lang.Thread").getDeclaredMethod("ensureMaterializedForStackWalk");
+        } catch (Exception e) {
+            return false;
+        }
+        return true;
+    }
+
+    private static void registerThreadPlugin(InvocationPlugins plugins, Replacements replacements) {
+        InvocationPlugin threadPlugin = new InvocationPlugin("ensureMaterializedForStackWalk", Object.class) {
+            @Override
+            public boolean apply(GraphBuilderContext b, ResolvedJavaMethod targetMethod, Receiver receiver, ValueNode object) {
+                b.add(new BlackholeNode(object));
```

Comment:
I think it would be helpful to add an extra `String` property to `BlackholeNode` so you can pass `"Thread.ensureMaterializedForStackWalk"` here and make it show up in IGV as a node labeled `Blackhole Thread.ensureMaterializedForStackWalk`. Otherwise people seeing a random blackhole in a graph might be very confused where it comes from, since this is a new use case for this node.

Reply:
Do I understand correctly that the following would suffice?

```diff
modified   compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/debug/BlackholeNode.java
@@ -51,12 +51,18 @@ public final class BlackholeNode extends FixedWithNextNode implements LIRLowerab
 
     public static final NodeClass<BlackholeNode> TYPE = NodeClass.create(BlackholeNode.class);
     @Input ValueNode value;
+    @OptionalInput String comment;
 
     public BlackholeNode(ValueNode value) {
         super(TYPE, StampFactory.forVoid());
         this.value = value;
     }
 
+    public BlackholeNode(ValueNode value, String comment) {
+        this(value);
+        this.comment = comment;
+    }
+
     public ValueNode getValue() {
         return value;
     }
modified   compiler/src/org.graalvm.compiler.replacements/src/org/graalvm/compiler/replacements/StandardGraphBuilderPlugins.java
@@ -2297,7 +2297,7 @@ public class StandardGraphBuilderPlugins {
         InvocationPlugin threadPlugin = new InvocationPlugin("ensureMaterializedForStackWalk", Object.class) {
             @Override
             public boolean apply(GraphBuilderContext b, ResolvedJavaMethod targetMethod, Receiver receiver, ValueNode object) {
-                b.add(new BlackholeNode(object));
+                b.add(new BlackholeNode(object, "ensureMaterializedForStackWalk"));
                 return true;
             }
         };
```

Reply:
Yes, that's about it. You have to remove the `@OptionalInput` on the `String`, only nodes can (and must) be inputs. Other properties are not annotated.
You should also add something like `nameTemplate = "Blackhole {p#reason}"` to the class's `@NodeInfo` so that the reason is shown as part of the name of the node itself. This might require `reason` to be non-`null`, so probably it should have a default value of `""`.